### PR TITLE
Add rollback_on_exit option

### PR DESCRIPTION
**Pull request by jmencak**
Add rollback option to tuned-main.conf file.  The option specifies how
TuneD should perform rollbacks.  By default, the old behaviour is
preserved (rollback=auto) and settings are rolled back on TuneD daemon
exit and profile switches.  Using rollback=not_on_exit will result in
TuneD not performing any rollbacks on the daemon exit, however, on
profile switches, settings rollbacks will still be performed.

Resolves: rhbz#2203142

### CODE REVIEW

**Comment by jmencak** [comment]
So, this is very much a WiP.  In order to actually prevent the rollback/unapplication of tuning, I either need to modify `instance_unapply_tuning()` in `tuned/plugins/base.py` or `_instance_unapply_static()` in every plug-in.  Obviously, the former is preferable, but I'm a bit confused about what `full_rollback` is intended to do.  @yarda , can we use that variable or do we need a new one for what I'm trying to do?  It is likely I probably need to rename the `rollback_on_exit` to something like `unapply_tuning_on_exit` and use it in `instance_unapply_tuning()` in `tuned/plugins/base.py`, right?  Thanks!

File: tuned/daemon/daemon.py
Comment (yarda) [inline tuned/daemon/daemon.py:182]: It also returns `True` if we don't know (or more precisely if `systemd` doesn't know or is misconfigured/in the failed state, etc.), thus maybe rename the method to "_system_shutdown_or_reboot_or_unknown" or similarly to clearly indicate the condition.

**Review by yarda — CHANGES_REQUESTED** [review]

**Comment by yarda** [comment]
Thanks. I am now running some tests, but it mostly LGTM.

I initially thought about optimization not to rollback settings that will be changed consequently, but it would require deeper TuneD engine changes. We could of course implement it later.

The proposed PR is more like hack, but it's simple, it shouldn't break anything and if it works for you, let's go with it for now.

File: tuned/daemon/daemon.py
Comment (jmencak) [reply tuned/daemon/daemon.py]: I wasn't able to come up with a function name for this I liked, so I reverted the function name to what it was originally and also reverted the logic in the fuction.

**Comment by jmencak** [comment]
> Thanks. I am now running some tests, but it mostly LGTM.
> 
> I initially thought about optimization not to rollback settings that will be changed consequently, but it would require deeper TuneD engine changes. We could of course implement it later.
> 
> The proposed PR is more like hack, but it's simple, it shouldn't break anything and if it works for you, let's go with it for now.

Would the future rewrite take advantage of the new `rollback_on_exit` function?  If not, I wouldn't like to add tech debt to TuneD.  Also, let's have another pair of eyes on this (Martin) if he has time.

File: tuned/daemon/daemon.py
Comment (MarSik) [inline tuned/daemon/daemon.py:232]: I think this need a bit more verbose comment, because the if right bellow says the same thing and sets full_rollback. The naming is a bit confusing here.

**Comment by yarda** [comment]
> > Thanks. I am now running some tests, but it mostly LGTM.
> > I initially thought about optimization not to rollback settings that will be changed consequently, but it would require deeper TuneD engine changes. We could of course implement it later.
> > The proposed PR is more like hack, but it's simple, it shouldn't break anything and if it works for you, let's go with it for now.
> 
> Would the future rewrite take advantage of the new `rollback_on_exit` function? If not, I wouldn't like to add tech debt to TuneD. Also, let's have another pair of eyes on this (Martin) if he has time.

If the RFE would ever happen we will keep this functionality. I think it shouldn't break anything and there could be people who do not have any interest in the rollback feature, thus having the configuration switch for it makes perfect sense to me.

For the RFE the idea is to "preload" the new profile and check whether there is an intersection of some plugins/commands. If it is, just skip the rollback for such command and instead of reading the initial value in the constructor, reuse the original value. This should be straightforward for the device/non-device class commands. But it could be tricky for the custom class commands which could do anything imaginable.

File: tuned-main.conf
Comment (yarda) [inline tuned-main.conf:80]: Maybe for future extensibility use `rollback` option name and string values e.g. `on_exit` and `auto` with the default value `auto`, e.g.:
```
rollback = on_exit
```
And we could add later more options like e.g.: `never`, `always`, etc.

We should also add the `tuned-main.conf` description to the upstream documentation, but this could be done later. It seems there is no dedicated section in the documentation yet.

Comment (jmencak) [reply tuned-main.conf]: Thank you for the suggestion.  I believe we need "auto" and "not_on_exit" right now.  Code adjusted.

File: tuned/daemon/daemon.py
Comment (jmencak) [reply tuned/daemon/daemon.py]: Comments added.

**Review by yarda — APPROVED** [review]

**Comment by yarda** [comment]
> LGTM, waiting for @MarSik comments.

Just two minor things I noticed:
- Maybe the `soft_rollback` is a bit confusing name, because IMHO with the `soft_rollback == False` no rollback would happen at all. Maybe you could indicate the no rollback with the `full_rollback = None` (quite hackish) or instead of the `full_rollback` and `soft_rollback` use just `rollback` with e.g. numbers (e.g. 0 - no rollback, 1 - normal rollback, 2 - full rollback - we shouldn't use the enum type, because we didn't drop python 2).
- Maybe add more verbose comment to the `tuned-main.conf`.

**Comment by jmencak** [comment]
Hopefully, all comments are addressed now.  I've kept the `full_rollback` variable.  Perhaps, we should rename it to `rollback_type` or similar in the future or perhaps even in this PR if you feel it should be done now.  The semantics are as proposed by Jarda:

```
full_rollback: 0 # no rollback
full_rollback: 1 # normal rollback (settings kept in volatile memory/RAM by TuneD are reverted)
full_rollback: 2 # full rollback (persistent changes done on the filesystem by TuneD are reverted)
```

This is a more intrusive change than what I had before when I added `soft_rollback`. Hopefully, I did not miss anything.  I'll run NTO e2e tests with this and report back.  Running beaker tests with this would also help.

File: tuned/daemon/daemon.py
Comment (jmencak) [reply tuned/daemon/daemon.py]: Thank you, bot, fixed.

File: tuned-main.conf
Comment (MarSik) [inline tuned-main.conf:81]:
```suggestion
#   graceful TuneD process exit
```

File: tuned/plugins/base.py
Comment (MarSik) [inline tuned/plugins/base.py:305]: Will this break profile change? Because we want to make sure that sysctls mentioned in the profile being unloaded are reverted. The new profile might not be touching them.

Comment (MarSik) [reply tuned/plugins/base.py:305]: But since it is a parameter, it probably gets decided elsewhere and here it just selects between soft (in-memory) and full rollback, right?

Comment (jmencak) [reply tuned/plugins/base.py:305]: > Will this break profile change? 

No

> Because we want to make sure that sysctls mentioned in the profile being unloaded are reverted.

They will be.  During profile change, `stop_tuning()` will be called with "full_rollback == 2".  However, on `TuneD` process termination, it will be called with "full_rollback == 0" if the user specified "rollback = not_on_exit".

> But since it is a parameter, it probably gets decided elsewhere

Yes.

File: tuned-main.conf
Comment (jmencak) [reply tuned-main.conf]: Thank you, fixed.

**Comment by yarda** [comment]
> This is a more intrusive change than what I had before when I added `soft_rollback`. Hopefully, I did not miss anything. I'll run NTO e2e tests with this and report back. Running beaker tests with this would also help.

Yep, it is, but I think it's much cleaner now.

Only nit-picking, but maybe you could use consts and not immediate int literals for profile rollback mode, well it will be still C-ish approach, but I think the code would be cleaner.

**Comment by jmencak** [comment]
> Only nit-picking, but maybe you could use consts and not immediate int literals for profile rollback mode, well it will be still C-ish approach, but I think the code would be cleaner.

Done.  As for my testing on OCP, some of the e2e tests are failing, but that' because they rely on the original rollback functionality.  I'll exclude them and rerun.

**Comment by jmencak** [comment]
All relevant e2e NTO tests with "rollback = not_on_exit" are passing now.

**Comment by yarda** [comment]
Well I think the variable `full_rollback` should be now called just `rollback` and if I am not mistaken the `consts.ROLLBACK_NOT_ON_EXIT` should be probably called `consts.ROLLBACK_NONE` or similarly because the upper logic decides if to do rollback or not and the underlying method doesn't care about the reason - it just get the argument don't do the rollback.

But it's just a cosmetic and I don't want to delay the release end enforce new re-testing, thus I will merge it as it is, but If you agree with my arguments please do another PR.

**Comment by jmencak** [comment]
Thank you for accepting this PR.

> Well I think the variable `full_rollback` should be now called just `rollback`

I intentionally kept `full_rollback` for easier search and replace.  For easier future search and replace, would you be OK with renaming it to `rollback_type`?

> and if I am not mistaken the `consts.ROLLBACK_NOT_ON_EXIT` should be probably called `consts.ROLLBACK_NONE` or similarly because the upper logic decides if to do rollback or not and the underlying method doesn't care about the reason - it just get the argument don't do the rollback.

I'm fine with this renaming.

> But it's just a cosmetic and I don't want to delay the release end enforce new re-testing, thus I will merge it as it is, but If you agree with my arguments please do another PR.

I'll definitely do this once we agree on the naming.  Thank you.

**Pull request merged**
